pennylane/capture/autograph/ag_primitives.py

```diff
@@ -84,6 +94,90 @@ def functional_cond():
     set_state(results)
 
 
+def assert_iteration_inputs(inputs, symbol_names):
+    """All loop carried values, variables that are updated each iteration or accessed after the
+    loop terminates, need to be initialized prior to entering the loop.
+
+    The reason is two-fold:
+      - the type information from those variables is required for tracing
+      - we want to avoid access of a variable that is uninitialized, or uninitialized in a subset
+        of execution paths
+
+    Additionally, these types need to be valid JAX types.
+    """
+
+    for i, inp in enumerate(inputs):
+        if isinstance(inp, Undefined):
+            raise AutoGraphError(
+                f"The variable '{inp}' is potentially uninitialized:\n"
+                " - you may have forgotten to initialize it prior to accessing it inside a loop, or"
+                "\n"
+                " - you may be attempting to access a variable local to the body of a loop in an "
+                "outer scope.\n"
+                f"Please ensure '{inp}' is initialized with a value before entering the loop."
+            )
+
+        try:
+            jax.api_util.shaped_abstractify(inp)
+        except TypeError as e:
+            raise AutoGraphError(
+                f"The variable '{symbol_names[i]}' was initialized with type {type(inp)}, "
+                "which is not compatible with JAX. Typically, this is the case for non-numeric "
+                "values.\n"
+                "You may still use such a variable as a constant inside a loop, but it cannot "
+                "be updated from one iteration to the next, or accessed outside the loop scope "
+                "if it was defined inside of it."
+            ) from e
+
+
+def assert_iteration_results(inputs, outputs, symbol_names):
+    """The results of a for loop should have the identical type as the inputs, since they are
+    "passed" as inputs to the next iteration. A mismatch here may indicate that a loop carried
+    variable was initialized with wrong type.
+    """
+
+    for i, (inp, out) in enumerate(zip(inputs, outputs)):
+        inp_t, out_t = jax.api_util.shaped_abstractify(inp), jax.api_util.shaped_abstractify(out)
+        if inp_t.dtype != out_t.dtype or inp_t.shape != out_t.shape:
+            raise AutoGraphError(
+                f"The variable '{symbol_names[i]}' was initialized with the wrong type, or you may "
+                f"be trying to change its type from one iteration to the next. "
+                f"Expected: {out_t}, Got: {inp_t}"
+            )
+
+
+def _call_pennylane_while(loop_test, loop_body, get_state, set_state, symbol_names):
+    """Dispatch to a PennyLane implementation of while loops."""
+
+    init_iter_args = get_state()
+    assert_iteration_inputs(init_iter_args, symbol_names)
+
+    def test(state):
+        old = get_state()
+        set_state(state)
+        res = loop_test()
+        set_state(old)
```

Comment:
Is this to prevent the loop test from changing the value of any local variables? What if that was the intended behaviour?

Reply:
Interesting question 🤔 I've tried for a bit and can't come up with a way to make a loop test that changes a local variable (beyond inside the loop test). Can you be more specific? 

No tests fail if we remove getting and resetting the state, though, so I'm not convinced this is necessary. It's just copied from the catalyst implementation.

Reply:
I was thinking something along the lines of 
```
while (x := ...) == ...:
   ...
```
Not sure how this is handled.

Reply:
Hmm... I tried it with this:

```
def fn(limit):
    i = 0
    y = 0
    while (i := y) < limit:
        y += 1
    return i

ag_fn = run_autograph(fn)
jaxpr = jax.make_jaxpr(ag_fn)(0)
```
And I get:

```
>>> fn(10)
10
```
```
>>>jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 10)
[0]
```
with or without these lines. So maybe the autograph setup doesn't allow the loop test to modify variables anyway? I guess for now that's unsupported. I'll leave this as-is for now, and bring it up when I'm back on Tuesday in case we want to document it as a limitation somewhere. It would be a good thing to include in the sharp-bits documentation!